server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -519,7 +519,26 @@ private SnapshotsInRepo sortSnapshotsWithNoOffsetOrLimit(List<SnapshotInfo> snap
             return sortSnapshots(snapshotInfos.stream(), snapshotInfos.size(), 0, GetSnapshotsRequest.NO_LIMIT);
         }
 
-        private SnapshotsInRepo sortSnapshots(Stream<SnapshotInfo> infos, int totalCount, int offset, int size) {
+        private SnapshotsInRepo sortSnapshots(Stream<SnapshotInfo> snapshotInfoStream, int totalCount, int offset, int size) {
+            final var resultsStream = snapshotInfoStream.filter(buildAfterPredicate()).sorted(buildComparator()).skip(offset);
+            if (size == GetSnapshotsRequest.NO_LIMIT) {
+                return new SnapshotsInRepo(resultsStream.toList(), totalCount, 0);
+            } else {
+                final var results = new ArrayList<SnapshotInfo>(size);
+                var remaining = 0;
+                for (var iterator = resultsStream.iterator(); iterator.hasNext();) {
+                    final var snapshotInfo = iterator.next();
+                    if (results.size() < size) {
+                        results.add(snapshotInfo);
+                    } else {
+                        remaining += 1;
+                    }
+                }
```

Comment:
Not sure I follow. Could we do the same with `java.util.stream.Stream#limit`?

Reply:
I realize, we need remaining, but can it be computed as `Math.max(0, totalCount - offset - size)`?

Reply:
I don't think so, we call this twice and `totalCount` isn't accurate the second time round.